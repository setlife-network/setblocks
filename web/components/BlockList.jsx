import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { cloneDeep } from 'lodash';

import SetBlock from './SetBlock';
import Text from './Text';
import { DEFAULT_SETBLOCKS } from '../constants';

import {
    setEnableSubmit, updateUnsavedSetblocks,
} from '../reducers/environment';

class BlockList extends React.Component {

    completeWithEmptySetBlocks(setBlocks) {
        const defaultSetBlocks = cloneDeep(DEFAULT_SETBLOCKS);
        let replacedBlocks = 0;
        setBlocks = _.orderBy(setBlocks, ['blockTime'], ['asc']); // Use Lodash to sort array by 'name'
        setBlocks = _.uniqBy(setBlocks, 'blockTime'); // Use Lodash to delete repeated blockTimes
        if (setBlocks && setBlocks.length > 0) { // To avoid iterate if the array is empty, return default
            defaultSetBlocks.forEach((setBlock, index, theArray) => {
                if (setBlocks && replacedBlocks < setBlocks.length && setBlock.blockTime === setBlocks[replacedBlocks].blockTime) {
                    theArray[index] = setBlocks[replacedBlocks];
                    replacedBlocks++;
                }
            });
        }
        return defaultSetBlocks
    }

    updateUnsavedSetBlock(selectedDay, index, editedSetBlock ) {
        const { unsavedSetBlocks } = this.props
        let dayEdited = unsavedSetBlocks[selectedDay];
        dayEdited[index] = editedSetBlock; // index 0 = SetBlock with blockTime 1
        this.props.updateUnsavedSetblocks({
            ...unsavedSetBlocks,
            [selectedDay]: dayEdited
        })
        this.props.setEnableSubmit(true)
    }

    render() {
        const {
            editModeSchedule, currentWeeklySetblocks, unsavedSetBlocks, selectedDay 
        } = this.props;
        const selectedDayFormatted = moment(selectedDay).format('YYYY-MM-DD')
        let setBlocksByDate =  _.groupBy(currentWeeklySetblocks, 'date');
        let setBlocks = setBlocksByDate[selectedDayFormatted];

        if (editModeSchedule && unsavedSetBlocks ) {
            // As it is your schedule, you can see the empty blocks, to complete then, that's why it is completed with the missing ones
            return this.completeWithEmptySetBlocks(setBlocks).map((setBlock, index) => {
                return (
                    <SetBlock
                        data={setBlock}
                        key={setBlock.id || (index + selectedDayFormatted)}
                        editMode={editModeSchedule}
                        updateSetBlock={(editedSetBlock) => this.updateUnsavedSetBlock(selectedDayFormatted, index, editedSetBlock )}
                    />
                )
            })
        } else if (!editModeSchedule && setBlocks) {
            setBlocks = _.orderBy(setBlocks, 'blockTime') // To properly render in order
            return setBlocks.map((setBlock, index) => {
                return <SetBlock data={setBlock} key={setBlock.id || index} />
            })
        } else {
            return (
                <Text weight='600' align='center'> This user hasn't committed any Setblocks for this day </Text>
            )
        }
    }
}

const mapStateToProps = ({ environment }) => {
    return {
        ...environment
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        setEnableSubmit: (enableSubmit) => dispatch(setEnableSubmit(enableSubmit)),
        updateUnsavedSetblocks: (unsavedSetBlocks) => dispatch(updateUnsavedSetblocks(unsavedSetBlocks))
    };
};
export default connect(mapStateToProps, mapDispatchToProps)(BlockList);