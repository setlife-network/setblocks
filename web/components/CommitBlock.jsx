import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { Check } from 'styled-icons/feather/Check.cjs'

import Card from './Card';
import Flex from './Flex';
import Text from './Text';

import {
    createSetBlock, deleteSetblock, setEditModeSchedule, updateSetBlock 
} from '../reducers/environment';

class CommitBlock extends React.Component {

    state = {
        showToast: false
    }

    upsertSetBlocks = () => {
        const { selectedDay, unsavedSetBlocks, currentTeamMember, enableSubmit } = this.props
        const day = moment(selectedDay).format('YYYY-MM-DD');
        if (enableSubmit) {

            unsavedSetBlocks[day].map( (setBlock) => {
                if (setBlock.id) {
                // Update
                    if (setBlock.blockFraction !== 0) {
                    // Update if change description, issueUrl, or blockFraction
                        this.props.updateSetBlock(setBlock)
                    } else {
                    // Delete if blockFraction is 0
                        this.props.deleteSetblock({ setblockId: setBlock.id })
                    }
                } else if (setBlock.blockFraction !== 0) {
                // Create a new one if have blockFraction != 0
                    this.props.createSetBlock({
                        teamMemberId: currentTeamMember.id,
                        date: day,
                        ...setBlock,
                        issueUrl: (setBlock.issueUrl || '')
                    })
                }
            })

            this.setState({
                showToast: true // Improve toast system in a future
            })
            _.delay( () => { // This probably made a warning but is only temporal, this will be removed with the toast system
                this.setState({
                    showToast: false // Improve toast system in a future
                })
                this.props.setEditModeSchedule(false);
            }, 1000);
        }
    }

    countHours() {
        const { selectedDay, unsavedSetBlocks } = this.props
        const day = moment(selectedDay).format('YYYY-MM-DD');
        return unsavedSetBlocks[day]
            ? unsavedSetBlocks[day].reduce( (last, setBlock) => {
            // Math.abs is used because the second blockFraction is represented by -0.5
                return last + (Math.abs(setBlock.blockFraction) * 4)
            }, 0)
            : 0 // If not defined yet, return 0
    }

    countBlockFractions() {
        const { selectedDay, unsavedSetBlocks } = this.props
        const day = moment(selectedDay).format('YYYY-MM-DD');
        return unsavedSetBlocks[day]
            ? unsavedSetBlocks[day].reduce( (last, setBlock) => {
            // Math.abs is used because the second blockFraction is represented by -0.5
                return last + (Math.abs(setBlock.blockFraction))
            }, 0)
            : 0 // If not defined yet, return 0
    }

    render() {
        const { selectedDay, enableSubmit } = this.props
        const { showToast } = this.state
        const onSaveToast = (
            <Text color='accent'>
                {'SetBlock committed successfully!!'}
            </Text>
        )

        return (
            <Flex column center width='100%'>
                <Text
                    weight='600'
                    aling='center'
                    size='10px'
                    ml='-10rem'
                >
                    {this.countHours() + ' work hours scheduled'}
                </Text>

                <Flex row center>
                    {enableSubmit
                    && (
                        <Text
                            weight='600'
                            aling='center'
                            size='10px'
                            whiteSpace='pre'
                            mx='0.5rem'
                        >
                            {' Commit ' + this.countBlockFractions() + ' SetBlocks \n '
                        + 'for ' + moment(selectedDay).format('dddd, MMM Do') + '?'}
                        </Text>
                    )}
                    <Card
                        bg={enableSubmit ? '#F5F5F6' : 'overlay'}
                        minWidth='20%'
                        maxWidth='100px'
                        height='40px'
                        mx='auto'
                        borderRadius='5px'
                        depth={7}
                        onClick={() => this.upsertSetBlocks()}
                    >
                        <Text color='textSecondary' align='center' size='1rem' weight='600' mt='8px' mx='5px'>
                            {'Commit'}
                            {React.createElement(Check, {
                                size: 24,
                                color: 'black'
                            })}
                        </Text>
                    </Card>
                </Flex>
                {showToast && (onSaveToast)}
            </Flex>
        );
    }
}

const mapStateToProps = ({ environment }) => {
    return {
        ...environment
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        createSetBlock: (params) => dispatch(createSetBlock(params)),
        deleteSetblock: (params) => dispatch(deleteSetblock(params)),
        updateSetBlock: (params) => dispatch(updateSetBlock(params)),
        setEditModeSchedule: (editMode) => dispatch(setEditModeSchedule(editMode))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(CommitBlock);
