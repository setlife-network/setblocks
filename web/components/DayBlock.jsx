import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import Card from './Card';
import Flex from './Flex';
import Text from './Text';

import theme from '../styles/theme'

class DayBlock extends React.Component {

    renderTinySetBlocks = (setBlocks, day, fetchingData) => {
        const blockDay = moment(day).format('YYYY-MM-DD')
        const setBlocksByDate = _.groupBy(setBlocks, 'date')
        let setBlocksToRender = setBlocksByDate[blockDay] || [];
        setBlocksToRender = _.orderBy(setBlocksToRender, 'blockTime') // Order by blockTime to have all the tinyBlock ordered
        if (setBlocksToRender.length > 0 && !fetchingData) {
            return setBlocksToRender.map(setBlock => (
                <Card
                    key={setBlock.id}
                    height='8px'
                    width='8px'
                    borderBottom={setBlock.blockFraction === 1.0 ? `4px ${theme.colors.accent} solid` : (setBlock.blockFraction === 0.5 ? `4px ${theme.colors.accent} solid` : '')}
                    borderTop={setBlock.blockFraction === 1.0 ? `4px ${theme.colors.accent} solid` : (setBlock.blockFraction === -0.5 ? `4px ${theme.colors.accent} solid` : '')}
                    bg='primaryLight'
                    my='0.3rem'
                    mr='0.3rem'
                >
                </Card>
            ))
        } else {
            return <Card height='8px' width='5px' bg='none' my='0.3rem' mr='0.3rem' /> // To prevent the sidebar grows slightly in width when render the tinySetBlocks
        }
    }

    render() {
        const { day, selected, onClick, fetchingData, currentWeeklySetblocks } = this.props

        return (
            <Flex
                column
                center
                className='DayBlock'
                style={selected ? { borderBottom: `2px ${theme.colors.grey} solid`, borderTop: `2px ${theme.colors.grey} solid` } : {}}
            >
                <Card
                    width='100%'
                    bg={selected ? theme.colors.lightGrey : theme.colors.primary}
                    mx='0.5rem'
                    onClick={() => onClick(day)}
                >
                    <Flex row center>
                        <Flex column mx='0.5rem'>
                            { // If you are waiting for the API to respond, it does not render
                                this.renderTinySetBlocks(currentWeeklySetblocks, day, fetchingData)
                            }
                        </Flex>
                        <Flex column mr='auto'>
                            <Text align='center' mb='0rem' size={25} lineHeight='1' color={selected ? 'accent' : theme.colors.overlay}>
                                {day.getDate()}
                            </Text>
                            <Text align='center' mt='0rem' size={11} lineHeight='1' color={selected ? 'accent' : theme.colors.overlay}>
                                {day.toDateString().slice(0, 3)}
                            </Text>
                        </Flex>
                    </Flex>
                </Card>
            </Flex>
        );
    }
}

const mapStateToProps = ({ environment }) => {
    return {
        ...environment
    };
};

const mapDispatchToProps = {
};
export default connect(mapStateToProps, mapDispatchToProps)(DayBlock)