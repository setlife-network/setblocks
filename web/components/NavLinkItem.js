import React from 'react'
import { Link } from 'react-router-dom'

import Box from './Box'
import Flex from './Flex'
import Text from './Text'
import Card from './Card';

import theme from '../styles/theme'

export default ({
    text, selected, icon, route, width
}) => (
    <Box
        width={width + '%'}
    >
        <Link
            to={route}
            style={{ textDecoration: 'none' }}
        >
            <Card
                bg={theme.colors.charcoal}
                width='100%'
                height='100%'
                borderTop={selected ? `2px ${theme.colors.accent} solid` : ''}
                depth={8}
            >
                <Flex
                    column
                    height='100%'
                    py='5px'
                    alignItems='center'
                >
                    {React.createElement(icon, {
                        size: 24,
                        color: theme.colors.white
                    })}
                    <Text
                        color={theme.colors.white}
                        size={[10]}
                        my='0'
                        weight='600'
                    >
                        {text}
                    </Text>
                </Flex>
            </Card>
        </Link>
    </Box>
)
