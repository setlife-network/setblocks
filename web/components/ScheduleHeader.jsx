import React from 'react';
import moment from 'moment';

import Text from './Text';
import Card from './Card';
import Flex from './Flex';
import theme from '../styles/theme';


export default class ScheduleHeader extends React.Component {
    render() {
        const { selectedDay } = this.props;
        return (
            <Flex
                row
                className='ScheduleHeader'
                center
                width='100%'
            >
                <Card
                    width='10px'
                    height='10px'
                    bg={theme.colors.transparent /* For the moment */}
                    borderRadius={50}
                    my='auto'
                    mx='0.5rem'
                />
                <Text
                    size={27}
                    align='center'
                    color={theme.colors.textPrimary}
                    my='0.5rem'
                    mx='auto'
                >
                    {moment(selectedDay).format('dddd Do')}
                </Text>

            </Flex>
        );
    }
}
