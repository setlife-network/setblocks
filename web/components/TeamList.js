import React from 'react';
import { connect } from 'react-redux'

import TeamMember from './TeamMember'

import { fetchAllTeamMembers } from '../reducers/environment'

class TeamList extends React.Component {

    componentDidMount() {
        this.props.fetchAllTeamMembers()
    }

    generateColor () {
        return '#' + Math.random().toString(16).substr(-6);
    }

    render() {
        const { teamMembers, goToPage, goToEdit } = this.props;
        return (
            <>
                {teamMembers.map(teamMember => {
                    return (
                        <TeamMember
                            key={teamMember.id}
                            name={teamMember.name}
                            goToPage={() => { goToPage(teamMember) }}
                            goToEdit={() => { goToEdit(teamMember) }}
                            color={this.generateColor()}
                        />
                    )
                })
                }
            </>
        )
    }
}

const mapStateToProps = ({ environment }) => {
    return {
        ...environment
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        fetchAllTeamMembers: () => dispatch(fetchAllTeamMembers())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(TeamList)