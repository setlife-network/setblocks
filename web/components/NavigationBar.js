import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash'

import NavLinkItem from './NavLinkItem';

import { NAVIGATION } from '../constants';
import Flex from './Flex';

export class NavigationBar extends React.Component {
    renderNavItem = (item) => {
        const { path } = this.props;
        const route = '/' + _.split(path, '/', 2)[1];
        return (
            <NavLinkItem
                {...item}
                selected={item.route === route}
                width={100 / NAVIGATION.items.length}
            />
        )
    }

    render() {
        return (
            <Flex
                className='NavigationBar'
                width='100%'
            >
                <Flex
                    className='nav-container'
                    width='100vw'
                >
                    {NAVIGATION.items.map(this.renderNavItem)}
                </Flex>
            </Flex>
        );
    }
}

const mapStateToProps = (state) => {
    return {};
};

const mapDispatchToProps = (dispatch) => {
    return {};
};

export default connect(mapStateToProps, mapDispatchToProps)(NavigationBar);
