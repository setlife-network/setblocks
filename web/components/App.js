import React from 'react';
import { compose } from 'redux'
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { Elements, StripeProvider } from 'react-stripe-elements';

import { STRIPE_PUBLIC_KEY } from '../constants'

import Flex from './Flex'
import NavigationBar from './NavigationBar';

import routes from '../routes'

import { checkAuthentication } from '../reducers/auth'

class App extends React.Component {
    componentDidMount() {
        console.log('componentDidMount')
        // Run initialization functions here
        this.props.checkAuthentication()
        .then(loggedIn => {
            // if (loggedIn) this.props.history.push('/team')
        })
    }

    render() {
        const { location } = this.props;

        return (
            <>
                <StripeProvider apiKey={STRIPE_PUBLIC_KEY}>
                    <Flex height='92vh'>
                        <Elements>
                            {routes}
                        </Elements>
                    </Flex>
                </StripeProvider>

                <Flex
                    height='8vh'
                    bg='primary'
                >
                    {this.props.loggedInUser.id != null &&
                        <NavigationBar path={location.pathname} />
                    }
                </Flex>
            </>
        );
    }
}

const mapStateToProps = ({ auth }) => {
    return {
        ...auth
    };
};

const mapDispatchToProps = {
    checkAuthentication
};

export default compose(
    withRouter,
    connect(mapStateToProps, mapDispatchToProps)
)(App)