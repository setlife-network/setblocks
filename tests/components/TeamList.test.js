import React from 'react';
import { shallow } from 'enzyme';
import { TeamList } from '../../web/components/TeamList';

let renderedComponent;
let MOCK_TEAM_MEMBERS = [
    {
        id: '00000001',
        name: 'Team Member #1'
    }
]

let mock = () => jest.fn();
let mockFetchAllTeamMembers

describe('TeamList Component', () => {
    beforeEach(() => {
        mockFetchAllTeamMembers = jest.fn().mockResolvedValue(MOCK_TEAM_MEMBERS)

        renderedComponent = shallow(
            <TeamList
                teamMembers={[]}
                fetchAllTeamMembers={mockFetchAllTeamMembers}
                // goToPage={mockedFunction}
                // goToEdit={mockedFunction}
            />
        );
    });

    it('should render correctly', () => {
        expect(renderedComponent).toMatchSnapshot();
    })

    it('should fetchAllTeamMembers only once', () => {
        expect(mockFetchAllTeamMembers.mock.calls.length).toBe(1)
    })

    it('fetchAllTeamMembers should resolve an array of team members', () => {
        mockFetchAllTeamMembers.mock.results[0].value.then(resolvedValue => {
            expect(resolvedValue).toBe(MOCK_TEAM_MEMBERS)
        })
    })
});