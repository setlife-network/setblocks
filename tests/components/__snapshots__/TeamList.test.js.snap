// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamList Component should render correctly 1`] = `<Fragment />`;
